### PR TITLE
divide before multiply to avoid overflow

### PR DESCRIPTION
**Pull request by geektortoise**
Hello, 

I had an integer overflow in the getRegion() method of the TileManager.cc file with the following values.

width : 14240
height : 13395
My in.dataLength was 35363488 because the width*height*channel*bpc was greater than INT_LIMIT before dividing by 8.

So I add parenthesis in each location than I found.

Have a nice day !

### CODE REVIEW

**Comment by stweil** [comment]
Is `bpc` always a multiple of 8?

**Comment by geektortoise** [comment]
Good question.
For what I saw here (https://github.com/ruven/iipsrv/blob/d2fd749ba10f40f92d40882b06902267275a7002/src/TileManager.cc#L363), it seems we only have 8,16 or 32 bpc supported. 

Yes, it will not avoid a future buffer overflow because there is no limit checking in the code.
I'm not familiar with C++ so I didn't begin to implement it without any previous example to base my code on.

**Comment by ruven** [comment]
Thanks @geektortoise. Currently, yes, bits per pixel is only a multiple of 8, so your solution works (or at least raises the overflow limit). Maybe a longer-term solution would be to change dataLength from an int type to long? Also at the moment it's a signed int, whereas this value is only ever positive.

We could also do limit checking through the C++ INT_MAX macro or using one of the solutions listed here: [https://stackoverflow.com/questions/199333/how-to-detect-integer-overflow](https://stackoverflow.com/questions/199333/how-to-detect-integer-overflow)

By the way, how did you get a RawTile of this size? Are you trying to export a huge CVT image?

File: src/TileManager.cc [hunk 1]
```diff
@@ -130,23 +130,23 @@ void TileManager::crop( RawTile *ttt ){
...
-  int len = tw * th * ttt->channels * ttt->bpc/8;
+  int len = tw * th * ttt->channels * (ttt->bpc/8);
   unsigned char* buffer = (unsigned char*) malloc( len );
   unsigned char* src_ptr = (unsigned char*) memcpy( buffer, ttt->data, len );
   unsigned char* dst_ptr = (unsigned char*) ttt->data;
 
   // Copy one scanline at a time
   for( unsigned int i=0; i<ttt->height; i++ ){
-    len =  ttt->width * ttt->channels * ttt->bpc/8;
+    len =  ttt->width * ttt->channels * (ttt->bpc/8);
```
Comment (stweil) [inline src/TileManager.cc]: This assignment could be placed before the for loop.

Comment (ruven) [reply src/TileManager.cc]: Yes, that would definitely be more efficient!

**Comment by geektortoise** [comment]
"By the way, how did you get a RawTile of this size? Are you trying to export a huge CVT image?"
Yes. With cytomine, we sometimes want to extract huge subparts of the image.
I know I am testing the limits of iipsrv ^^'

**Comment by ruven** [comment]
Thanks, now merged including the optimization mentioned by @stweil

**Pull request merged**